### DISCUSSION TITLE
关于stringify的疑问

### DISCUSSION THREAD

**Discussion by constly** [ejoy/ant#111] [General]
我最近在使用datalist进行持久化存储，具体为:
1. 使用 serialize.stringify({}) 将table序列化为字符串
2. 使用 datalist.parse(str) 将字符串反序列化为table

我发现 serialize.stringify 没有处理 key 为number 的情况，测试代码如下
```
local test = {
    b = 2,
    grids = {
	[4] = {x = 1, y = 4},
	[5] = {x = 1, y = 4},
	["6"] = {x = 1, y = 3},
    },
    test = "a",
}
local serialize = import_package "ant.serialize"
print(serialize.stringify(test))
```
序列化后的输出为:
```
b: 2
grids:
  6:
    x: 1
    y: 3
name: a
```
即grids下的4 和 5被忽略了。我看了下stringify.lua，发现sortpairs函数里面没有处理type为number的情况。
我又测试了下
```
local test = {
    b = 2,
    grids = {
	[5] = {x = 1, y = 4},
	["6"] = {x = 1, y = 3},
   },
   test = "a",
}
local serialize = import_package "ant.serialize"
print(serialize.stringify(test))
```
输出为:
```
b: 2
grids: {}
test: a
```
这种情况下，grids下的两个条目都被忽略。
总结下就是:
1. serialize.stringify 没有处理key为number的情况
2. 当数组中有key为number时，serialize.stringify 功能会出现异常，见上面第二个测试

也就是说，stringify 只支持table类型为数组 或者 key为string的情况，如果有key类型为number会出问题，这是bug还是设计如此呢~

**Comment by actboy168** [comment]
因为这本身就不是一个合法的datalist数据。datalist只是描述了lua的数据类型的一个子集，你要先考虑你的数据在datalist该如何表示，然后再转成相应的lua数据。

如果你只是为了序列化，可以用`require "bee.serialization"`,他可以提供更高的适用范围，但他是二进制格式的。

**Reply by constly** [reply]
好吧，不过我还是觉得datalist作为一个官方的编辑器下的序列化方案，如果不支持{10 = "a"} 这种情况，会有点奇怪，因为我觉得这种使用情形在lua中还是比较普遍的

**Reply by cloudwu** [reply]
绝大部分其它 标记语言 的字典类型都不支持 string 以外类型的 key 。例如 json yaml xml 等等。
如果 datalist 支持了，以后会难以将数据转换为其它标记语言表示。

另一方面，既然现存的各种标记语言都要求字典的 key 必须是 string ，说明这个限制是合理的。

**Reply by constly** [reply]
了解了，原来你们是这么来考虑的~

**Reply by cloudwu** [reply]
另外，datalist 可以写成
```
data:
  1000 : 1
  2000 : 2
```
只不过，这里的 1000 和 2000 解析后是字符串类型，1 和 2 是数字类型。
